18F/identity-idp
# LG-13963: 👤 Socure shadow mode background job

(Previously: #11093)

## 🎫 Ticket

Link to the relevant ticket:
[LG-13963](https://cm-jira.usa.gov/browse/LG-13963)

## 🛠 Summary of changes

Adds a new **disabled** background job, `SocureShadowModeProofingJob`, to do proofing with Socure KYC and log the result.

To enable the job, you need to set `idv_socure_shadow_mode: true` in your config. Then it will be scheduled automatically by the ResolutionProofingJob.

## 📜 Testing Plan

To test this locally, you will need a Socure KYC API key. To get one:

1. Log into the Socure dashboard
2. Make sure you're in the "Sandbox" environment
3. Navigate to the "ID+ Keys" tab in the "Developers" section and copy a key

Then update your `application.yml`:

```yaml
  idv_socure_shadow_mode_enabled: true
  socure_idplus_api_key: 'YOUR_API_KEY_HERE'
  socure_idplus_base_url: 'https://sandbox.socure.us'
  socure_idplus_timeout_in_seconds: 5
```
Then, complete identity verification. **DO NOT USE REAL PII.** You should see a new event logged in your events.log called `idv_socure_shadow_mode_proofing_result`:

<details>

<summary>Here is an example</summary>

```json
{
  "name": "idv_socure_shadow_mode_proofing_result",
  "properties": {
    "event_properties": {
      "resolution_result": {
        "success": true,
        "errors": {},
        "exception": null,
        "timed_out": false,
        "threatmetrix_review_status": "pass",
        "context": {
          "device_profiling_adjudication_reason": "device_profiling_result_pass",
          "resolution_adjudication_reason": "pass_resolution_and_state_id",
          "should_proof_state_id": true,
          "stages": {
            "resolution": {
              "success": true,
              "errors": {},
              "exception": null,
              "timed_out": false,
              "transaction_id": "resolution-mock-transaction-id-123",
              "reference": "aaa-bbb-ccc",
              "can_pass_with_additional_verification": false,
              "attributes_requiring_additional_verification": [],
              "vendor_name": "ResolutionMock",
              "vendor_workflow": null,
              "verified_attributes": null
            },
            "residential_address": {
              "success": true,
              "errors": {},
              "exception": null,
              "timed_out": false,
              "transaction_id": "",
              "reference": "",
              "can_pass_with_additional_verification": false,
              "attributes_requiring_additional_verification": [],
              "vendor_name": "ResidentialAddressNotRequired",
              "vendor_workflow": null,
              "verified_attributes": null
            },
            "state_id": {
              "success": true,
              "errors": {},
              "exception": null,
              "mva_exception": null,
              "requested_attributes": {},
              "timed_out": false,
              "transaction_id": "state-id-mock-transaction-id-456",
              "vendor_name": "StateIdMock",
              "verified_attributes": []
            },
            "threatmetrix": {
              "client": null,
              "success": true,
              "errors": {},
              "exception": null,
              "timed_out": false,
              "transaction_id": "ddp-mock-transaction-id-123",
              "review_status": "pass"
            }
          }
        },
        "ssn_is_unique": true
      },
      "socure_result": {
        "success": true,
        "errors": {
          "reason_codes": [
            "I919"
          ]
        },
        "exception": null,
        "timed_out": false,
        "transaction_id": "4fc4b744-79d0-4df7-8751-e087609fb1fa",
        "reference": "",
        "can_pass_with_additional_verification": false,
        "attributes_requiring_additional_verification": [],
        "vendor_name": "socure_kyc",
        "vendor_workflow": null,
        "verified_attributes": [
          "address",
          "first_name",
          "last_name",
          "ssn",
          "dob"
        ]
      }
    },
    "new_event": true,
    "path": null,
    "session_duration": null,
    "user_id": "96d74912-6af8-4106-8d93-7485f04a3137",
    "locale": "en"
  },
  "time": "2024-08-22T19:25:45.678Z",
  "id": "d367752a-707e-4708-8ac1-781167163b0c",
  "visitor_id": "0b2d3739-287f-47ef-90c4-952b2fedb72e",
  "visit_id": "8a3f78d2-406e-4481-b839-ac18c03c9fda",
  "log_filename": "events.log"
}
```

</details>

You should also verify that this job _does not_ run when `idv_socure_shadow_mode_enabled` is not set or is disabled.
